### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by snork-nbl**

### CODE REVIEW

File: lib/Auth.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    // Returns:                 Promise that resolves if operation succeeds, or rejects 
+    //                          with an error
+    getAccessToken(loginKey) {
+    }
+
+    // Authenticate a user via an identifier/password pair.
+    //
+    // If the authentication is successfully completed, the obtained access token to be automatically used
+    // by the library for further access to Imp Central API.
+    // Also, the access token and the obtained refresh token are returned to the caller in the HTTP response body.
```
Comment (ppetrosh) [inline lib/Auth.js]: Where is the response body returned to?

`getAccessToken(loginKey)` is the retrieval of access token by _loginKey_
`login(id, password)` is the retrieval of access token by _login/password_

And these two methods are completely orthogonal to each other.

Should they both be called getAccessToken...? just trying to better understand.

File: lib/DeviceGroups.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    //     pageSize   : Number  Optional pagination size - maximum number of items to return.
+    //                          If not specified, the default value is 20.
+    // 
+    // Returns:                 Promise that resolves when the Device Groups list is successfully
+    //                          obtained, or rejects with an error
+    list(filters = null, pageNumber = null, pageSize = null) {
+    }
+
+    // Possible filters keys for the list() filters:
+    static get FILTER_OWNER_ID() {
```
Comment (ppetrosh) [inline lib/DeviceGroups.js]: Not sure I understand this part. Why do we need this filter? How would someone retrieve it's (and even others') id?

File: lib/Devices.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    // Retrieves a specific Device.
+    //
+    // Parameters:
+    //     deviceId : String    ID of the Device to be retrieved
+    //
+    // Returns:                 Promise that resolves when the Device is successfully obtained, 
+    //                          or rejects with an error
+    get(deviceId) {
```
Comment (ppetrosh) [inline lib/Devices.js]: Is there anything else other than device name that is returned here?

**Comment by ppetrosh** [comment]
I was thinking of the following hierarchal APIs:

`ImpCentralAPI.getProduct("mySuperProduct").getDeviceGroup("myGroupName").getDevice("name or id").getName()`

or 

`ImpCentralAPI.getProduct("mySuperProduct").getDeviceGroup("myGroupName").getDevice("name or id").unassign()`

or 

`ImpCentralAPI.getProduct("mySuperProduct").deleteDeviceGroup("myGroupName")`

Did you consider this approach? What was the rationale behind this flatten structure?

**Comment by ppetrosh** [comment]
Another generic architectural question. Did you consider using callbacks vs async vs promises? Why do you think promise based API is the best choice here?

File: lib/Auth.js (hunk 1 above)
Comment (alexey-nbl) [inline lib/Auth.js]: There are 4 orthogonal ways to initialize the lib by an access token (or 5 if consider 2FA as a separate way) - described at the beginning of the file.
AFAIK, JS does not allow to have the methods with the same name.
We analyzed severall different naming approaches and suggest the current one.

**Comment by alexey-nbl** [comment]
About flatten structure:
1) Original v5 API has a flatten structure.
2) The goal of the lib is a wrapper for v5 API which provides JS interface. (But we may change the goal, if required).
3) Accounts and Auth are flatten.
4) Devices can belong not only to DeviceGroups but to be unassigned as well.
5) Product-DeviceGroup-Deployment can have hierarchal APIs. But I do not know if it would be possible to support such a line of calls - ImpCentralAPI.getProduct().deleteDeviceGroup(). etc.
Basing on all the above, I think a flatten structure is better in this particular lib.

File: lib/Auth.js (hunk 1 above)
Comment (alexey-nbl) [inline lib/Auth.js]: The response body is returned in Promise.

File: lib/DeviceGroups.js (hunk 2 above)
Comment (alexey-nbl) [inline lib/DeviceGroups.js]: v5 API has this filter. Why do not support it?
Owner's id is returned with almost every object returned by v5 API. User can filter other objects by this owner id.
I think, the right question here - how to get owner id of collabarators? It might be usefull. But it's a question to v5 API spec.

File: lib/Devices.js (hunk 3 above)
Comment (alexey-nbl) [inline lib/Devices.js]: Sure. Many attributes. Probably it makse sense to put a note in every method that the details of what is returned are in v5 API spec.

**Comment by alexey-nbl** [comment]
About callbacks vs async vs promises...

Callbacks is a nightmare for users. The lib is not for imp, so there is no reson to save footprint, so should be ok to use async/promises.
We do not know much about JS async - need to study, compare with promises and choose what is easy to use by user of the lib.

**Comment by ekaterina-nbl** [comment]
As I can see async functions are just syntactic sugar on top of Promises. For the current API I see no real advantages over promises.
Additionally it is a recent feature which is absent in Node.js v6.10.2 LTS recommended for most users.

About hierarchical API, I see no way to support a line of calls exactly like ImpCentralAPI.getProduct("mySuperProduct").getDeviceGroup("myGroupName").getDevice("name or id").unassign()
if all of the methods are asynchronous. Even if we use async functions.

In general, the lib user anyway will need to use something like promises chain :
ImpCentralAPI.getProduct("mySuperProduct")
    .then(product => product.getDeviceGroup("myGroupName"))
    .then(devGroup => devGroup.getDevice("name or id"))
    .then(device => device.unassign());

If the lib methods will be called from async function, it is possible to use awaits instead, but I'm not sure it's really simplified a code:
const product = await ImpCentralAPI.getProduct("mySuperProduct");
const devGroup = await product.getDeviceGroup("myGroupName");
const device = await devGroup.getDevice("name or id");
await device.unassign();

Could you please clarify how do you think to implement such a calls chains in hierarchical API?

File: lib/Auth.js (hunk 1 above)
Comment (ppetrosh) [inline lib/Auth.js]: may be `loginUsingOtp` -> `loginWithOTP`  or even `otpLogin`?

File: lib/DeviceGroups.js (hunk 2 above)
Comment (ppetrosh) [inline lib/DeviceGroups.js]: Got it. So if you are a collaborator on an account, your owner id will be seen there. Makes sense.

I believe you can query the user id by username/email (didn't check myself though).

File: lib/Devices.js (hunk 3 above)
Comment (ppetrosh) [inline lib/Devices.js]: Yep... with a link to the API spec?

File: lib/Auth.js (hunk 1 above)
Comment (alexey-nbl) [inline lib/Auth.js]: I agree with loginWithOTP

File: lib/DeviceGroups.js (hunk 2 above)
Comment (alexey-nbl) [inline lib/DeviceGroups.js]: The current v5 API allows only to query account by the Account's ID (or by "me")

File: lib/Devices.js (hunk 3 above)
Comment (alexey-nbl) [inline lib/Devices.js]: Maybe give an exact link only once - in ImpCentralApi.js ?
Or only in the readme file (not in source at all)?

Comment (ppetrosh) [inline lib/Devices.js]: makes sense

File: lib/DeviceGroups.js (hunk 2 above)
Comment (ppetrosh) [inline lib/DeviceGroups.js]: got it

File: README.md [hunk 4]
```diff
@@ -1 +1,279 @@
...
+
+**TODO: change the url in the example to the final one**
+
+```javascript
+const Imp = require('imp-central-api');
+const Errors = Imp.Errors;
+const imp = new Imp('https://api.ei.run/v5');
+
+let token;
+imp.auth.login(email, password).then(result => {
```
Comment (ppetrosh) [inline README.md]: I would probably define `email` and `password` as constants above with `"<user email for Electric Imp Account>"`, etc. to make it clear that it needs to be either provided by user.

File: README.md [hunk 5]
```diff
@@ -1 +1,279 @@
...
+To instantiate this library you need to call [ImpCentralApi class](./lib/ImpCentralApi.js) constructor.
+
+By default the library works with the following impCentral API base endpoint: **TODO: add the final link**
+
+You can optionally pass to the constructor an alternative impCentral API base endpoint. This can be used to connect to private impCloud installations.
+
+After instantiation use *ImpCentralApi* property getters to obtain the subclasses which provide methods to access impCentral API.
+
+```javascript
+const Imp = require('imp-central-api');
```
Comment (ppetrosh) [inline README.md]: It would probably be better to call the class `ImpCentralAPI` everywhere.

Comment (ppetrosh) [inline README.md]: `imp` is a not the best name overall as it means the specific hardware module + some marketing concept which we are highly encouraged not to abuse.

File: README.md [hunk 6]
```diff
@@ -1 +1,279 @@
...
+    }
+});
+```
+
+2. library initialization using existing access token, product and device group creation:
+
+```javascript
+const DeviceGroups = Imp.DeviceGroups;
+
+imp.auth.accessToken = token;
```
Comment (ppetrosh) [inline README.md]: Is it what users are supposed to do?

File: lib/util/HttpHelper.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+    static post(path, body = null) {
+        return HttpHelper.request('POST', path, null, body);
+    }
+
+    static patch(path, body) {
+        return HttpHelper.request('PATCH', path, null, body);
+    }
+
+    static delete(path, body = null) {
```
Comment (ppetrosh) [inline lib/util/HttpHelper.js]: seems like `delete` is a keyword that you might be overriding here

File: README.md (hunk 4 above)
Comment (ekaterina-nbl) [inline README.md]: done

File: README.md (hunk 5 above)
Comment (ekaterina-nbl) [inline README.md]: I renamed imp to impCentralApi and Imp to ImpCentralApi everywhere. Is it OK or did you mean something else?

File: README.md (hunk 6 above)
Comment (ekaterina-nbl) [inline README.md]: User can initialize the library this way if he already has a valid access token.
Alternatively he can initialize the library using any of the other ways described in the Authentication section.

File: lib/util/HttpHelper.js (hunk 7 above)
Comment (ekaterina-nbl) [inline lib/util/HttpHelper.js]: According to the ECMAScript 6 specification, delete is a reserved word which can not be used as a variable or function name, but can be used safely as a class methods or properties.
It doesn't override native delete operator.
But I can rename delete methods everywhere (almost all API classes like Products, DeviceGroups, etc contain it too) if you think it can cause any confusion.

Comment (ppetrosh) [inline lib/util/HttpHelper.js]: Got it. Thanks for clarification. It sounds like we are good here.

File: README.md (hunk 5 above)
Comment (ppetrosh) [inline README.md]: Yep, sounds good. Thank you.

File: README.md [hunk 8]
```diff
@@ -1 +1,280 @@
...
+    }
+});
+```
+
+2. library initialization using existing access token, product and device group creation:
+
+```javascript
+const DeviceGroups = ImpCentralApi.DeviceGroups;
+
+impCentralApi.auth.accessToken = token;
```
Comment (ppetrosh) [inline README.md]: It wasn't clear to me initially if one needs to instantiate the `impCentral` variable here. So probably it's worth to add some comments in the code or just create the instance on a line above.
